src/doc/tarpl/repr-rust.md

```diff
@@ -133,18 +132,20 @@ struct FooRepr {
 }
 ```
 
-And indeed this is approximately how it would be laid out in general
-(modulo the size and position of `tag`). However there are several cases where
-such a representation is inefficient. The classic case of this is Rust's
-"null pointer optimization". Given a pointer that is known to not be null
-(e.g. `&u32`), an enum can *store* a discriminant bit *inside* the pointer
-by using null as a special value. The net result is that
-`size_of::<Option<&T>>() == size_of::<&T>()`
+And indeed this is approximately how it would be laid out in general (modulo the
+size and position of `tag`). 
+
+However there are several cases where such a representation is inefficient. The
+classic case of this is Rust's "null pointer optimization": an enum consisting
+of a unit variant and a non-nullable pointer variant (e.g. `&u32`) makes the tag
```

Comment:
This isn't quite strong enough -- Rust can dig deep into types to find that pointer. For instance `Option<Vec<T>>` also optimizes. However to my knowledge it can't dig through other enums or use multiple nullable pointers -- so `Option<Option<(&u8, &u8)>>` is 24-bytes on 64 bit.

Also "unit variant" strikes me as a bit jargony, though I have no great concrete proposal -- perhaps introduce the notion of a C-like variant?